File: java/smartlist/src/main/java/com/smartlist/module/question/questionstrategy/McqSingleWeightedStrategy.java
package com.smartlist.module.question.questionstrategy;

import com.smartlist.module.question.QuestionResponseCodes;
import com.smartlist.module.question.dto.request.AnswerOptionReqDTO;
import com.smartlist.module.question.dto.request.QuestionReqDTO;
import com.smartlist.module.question.interfaces.IQuestionStrategy;

public class McqSingleWeightedStrategy implements IQuestionStrategy {

    private final QuestionReqDTO questionReqDto;

    public McqSingleWeightedStrategy(final QuestionReqDTO questionReqDto){
        this.questionReqDto = questionReqDto;
    }

    @Override
    public void validate() {
        int correctCount = this.questionReqDto.getAnswerOptions().stream().mapToInt(ao -> Boolean.TRUE.equals(ao.getCorrect()) ? 1 : 0).sum();
        if (correctCount != this.questionReqDto.getAnswerOptions().size()) {
            throw QuestionResponseCodes.INVALID_ANSWER_OPTION;
        }

        this.questionReqDto.getAnswerOptions().forEach(answerOptionDto -> {
            if (answerOptionDto.getScore() == null || answerOptionDto.getScore() < 0) {
                throw QuestionResponseCodes.INVALID_SCORE;
            }
        });

        int maxScore = 0;
        for(int i =0; i < this.questionReqDto.getAnswerOptions().size(); i++){
            AnswerOptionReqDTO answerOptionReqDto = this.questionReqDto.getAnswerOptions().get(i);
            if (answerOptionReqDto.getScore() > maxScore) {
                maxScore = answerOptionReqDto.getScore();
            }
        }
        this.questionReqDto.setScore(maxScore);
    }
}

File: javascript/smartlist/smartlist-api-mongodb-project/src/modules/question/question-strategy/mcq-single-weighted-strategy.ts
import { IQuestionStrategy } from '../question-interface/question-interface';
import { QuestionReqDto } from '../dto/request/question.req.dto';
import { QuestionResponseCodes } from '../question.response.codes';

export class McqSingleWeightedStrategy implements IQuestionStrategy {
  constructor(protected questionReqDto: QuestionReqDto) {}
  validate() {
    if (this.questionReqDto.answer_options.length < 2) {
      throw QuestionResponseCodes.INVALID_ANSWER_OPTIONS_FOR_QUESTION_TYPE;
    }

    let correctCount = 0;
    this.questionReqDto.answer_options.forEach((ao) => {
      correctCount += ao.correct ? 1 : 0;
    });
    if (correctCount != this.questionReqDto.answer_options.length) {
      throw QuestionResponseCodes.INVALID_ANSWER_OPTION;
    }

    this.questionReqDto.answer_options.forEach((answerOptionDto) => {
      if (answerOptionDto.score == null || answerOptionDto.score < 0) {
        throw QuestionResponseCodes.INVALID_SCORE;
      }
    });

    let maxScore = 0;
    this.questionReqDto.answer_options.forEach((answerOptionDto) => {
      if (Number(answerOptionDto.score) > maxScore) {
        maxScore = Number(answerOptionDto.score);
      }
    });
    this.questionReqDto.score = maxScore;
  }
}
